src/index/base.cpp

```diff
@@ -359,7 +359,10 @@ bool BaseIndex::Start()
     // Need to register this ValidationInterface before running Init(), so that
     // callbacks are not missed if Init sets m_synced to true.
     RegisterValidationInterface(this);
-    if (!Init()) {
+    if (!Init()) return false;
+
+    const CBlockIndex* index = m_best_block_index.load();
+    if (!CustomInit(index ? std::make_optional(interfaces::BlockKey{index->GetBlockHash(), index->nHeight}) : std::nullopt)) {
```

Comment:
In commit "indexes, refactor: Remove CBlockIndex* uses in index Init methods" (bef4e405f3de2718dfee279a9abff4daf016da26)

Calling the `CustomInit()` function after the `Init()` function here seems to cause a race condition, noticed by @furszy and described in https://github.com/bitcoin/bitcoin/pull/27607#issuecomment-1553400917. The problem is that `Init()` can set `m_synced` to true, which enables `BlockConnected` callbacks. But if `BlockConnected` callback runs before `CustomInit` finishes the index may still be only partially loaded, and the `CustomAppend` call in `BlockConnected` could fail